### ISSUE TITLE
libtorch: collate_fn equivalent

### ISSUE THREAD

**Issue by benedikth** [pytorch/pytorch#65984]
## 🚀 Feature
In PyTorch one can use `collate_fn` of/in a `DataLoader` in order to modify the batch sampling of the DataLoader. As far as I know, this is not possible in the C++-API / libtorch.

## Motivation

In order to support different dimensional input (like in my case images) this would be helpfull, because:

* One could apply some preprocessing steps like padding or cropping
* Or use the conv-layers of the CNN/NN in question performing a feature extraction resulting in an equally sized batch samples.

Right now there will occur some error in the batch sampling process (stack will throw an error because of different sized tensors).

## Pitch

It would be ideal, if it would be possible to

* Either define some `collate_fn` function like in PyTorch, or
* Get some tutorial-like instructions how to do this correctly.

## Alternatives

In the end I ended up overriding `get_batch(at::ArrayRef<size_t> indices)` ([stackoverflow-post](https://stackoverflow.com/questions/69358675/how-to-use-collate-fn-in-libtorch/69372321#69372321)) which is not ideal causing an error if the number of the dataLoders workers is greater than 0.

**Comment by ejguan** [comment]
In my opinion, firstly `collate_fn` is not supposed to be the part to handle data preprocessing. The goal of `collate_fn` is just stacking Tensors as a batch Tensor. The right way at least based on the normal workflow as torchvision, it should be implementing or using transform in your Dataset and make sure the images are with the same size.
